src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -0,0 +1,209 @@
+  private static Schema outputSchema;
+  private static String datasetProject;
+  private final BigQueryAvroToStructuredTransformer transformer = new BigQueryAvroToStructuredTransformer();
+  private final DataplexBatchSourceConfig config;
+  DataplexInterface dataplexInterface = new DataplexInterfaceImpl();
+
+  public DataplexBatchSource(DataplexBatchSourceConfig dataplexBatchSourceConfig) {
+    this.config = dataplexBatchSourceConfig;
+  }
+
+  @Override
+  public void transform(KeyValue<Object, Object> input, Emitter<StructuredRecord> emitter) throws IOException {
+    if (entityBean.getSystem().equalsIgnoreCase(BIGQUERY_DATASET_ENTITY_TYPE)) {
+      StructuredRecord transformed = outputSchema == null ?
+        transformer.transform((GenericData.Record) input.getValue()) :
+        transformer.transform((GenericData.Record) input.getValue(), outputSchema);
+      emitter.emit(transformed);
+    } else {
+      emitter.emit((StructuredRecord) input.getValue());
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    StageConfigurer configurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = configurer.getFailureCollector();
+    if (!config.getConnection().canConnect() || config.getServiceAccountType() == null ||
+      (config.isServiceAccountFilePath() && config.autoServiceAccountUnavailable()) ||
+      (config.tryGetProject() == null)) {
+      // ValidatingInputFormat plugin setup is mandatory.Otherwise pipeline will fail at runtime in case of GCS entity.
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, null);
+      return;
+    }
+    config.validateServiceAccount(collector);
+    entityBean = config.getAndValidateEntityConfiguration(collector, dataplexInterface);
+    if (entityBean == null) {
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, null);
+      return;
+    }
+    if (entityBean.getSystem().equals(BIGQUERY_DATASET_ENTITY_TYPE)) {
+      getEntityValuesFromDataPathForBQEntities(entityBean.getDataPath());
+      config.validateBigQueryDataset(collector, datasetProject, dataset, tableId);
+      config.validateTable(collector, datasetProject, dataset, tableId);
+      Schema configuredSchema = getOutputSchemaForBQEntity(collector);
+      configurer.setOutputSchema(configuredSchema);
+    } else {
+      config.checkMetastoreForGCSEntity(dataplexInterface, collector);
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, entityBean);
+    }
+  }
+
+  public void getEntityValuesFromDataPathForBQEntities(String dataPath) {
+    String[] entityValues = dataPath.split("/");
+    dataset = entityValues[entityValues.length - 3];
+    datasetProject = entityValues[1];
+    tableId = entityValues[entityValues.length - 1];
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+
```

Comment:
any validations done in configurePipeline should also be done in prepareRun.  I guess you're planning to add that later?

Reply:
yes, will add that code later.